src/textual/widgets/_tree.py

```diff
@@ -0,0 +1,838 @@
+        return first_line
+
+    def _add_node(
+        self,
+        parent: TreeNode[TreeDataType] | None,
+        label: Text,
+        data: TreeDataType | None,
+        expand: bool = False,
+    ) -> TreeNode[TreeDataType]:
+        node = TreeNode(self, parent, self._new_id(), label, data, expanded=expand)
+        self._nodes[node._id] = node
+        self._updates += 1
+        return node
+
+    def render_label(
+        self, node: TreeNode[TreeDataType], base_style: Style, style: Style
+    ) -> Text:
+        """Render a label for the given node. Override this to modify how labels are rendered.
+
+        Args:
+            node (TreeNode[TreeDataType]): A tree node.
+
+        Returns:
+            Text: A Rich Text object containing the label.
+        """
+        node_label = node._label.copy()
+        node_label.stylize(style)
+
+        if node._allow_expand:
+            prefix = (
+                "▼ " if node.is_expanded else "▶ ",
+                base_style + TOGGLE_STYLE,
+            )
+        else:
+            prefix = ("", base_style)
+
+        text = Text.assemble(prefix, node_label)
+        return text
+
+    def get_label_width(self, node: TreeNode[TreeDataType]) -> int:
+        """Get the width of the nodes label.
+
+        The default behavior is to call `render_node` and return the cell length. This method may be
+        overridden in a sub-class if it can be done more efficiently.
+
+        Args:
+            node (TreeNode[TreeDataType]): A node.
+
+        Returns:
+            int: Width in cells.
+        """
+        label = self.render_label(node, NULL_STYLE, NULL_STYLE)
+        return label.cell_len
+
+    def clear(self) -> None:
+        """Clear all nodes under root."""
+        self._tree_lines_cached = None
+        self._current_id = 0
+        root_label = self.root._label
+        root_data = self.root.data
+        self.root = TreeNode(
+            self,
+            None,
+            self._new_id(),
+            root_label,
+            root_data,
+            expanded=True,
+        )
+        self._updates += 1
+        self.refresh()
+
+    def select_node(self, node: TreeNode | None) -> None:
+        """Move the cursor to the given node, or reset cursor.
+
+        Args:
+            node (TreeNode | None): A tree node, or None to reset cursor.
+        """
+        self.cursor_line = -1 if node is None else node._line
+
+    def get_node_at_line(self, line_no: int) -> TreeNode[TreeDataType] | None:
+        """Get the node for a given line.
+
+        Args:
+            line_no (int): A line number.
+
+        Returns:
+            TreeNode[TreeDataType] | None: A tree node, or ``None`` if there is no node at that line.
+        """
+        try:
+            line = self._tree_lines[line_no]
+        except IndexError:
+            return None
+        else:
+            return line.node
+
+    def validate_cursor_line(self, value: int) -> int:
```

Comment:
Docstring?

Reply:
Good call; I've been on and off wondering about this too. I guess in the end too much documentation is better than none.